---
title: "Predictive Survival"
subtitle: "Metastatic cohort, drugs grouped by class"
author: "Alex Paynter"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: paper
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, include = F,
                      fig.width = 7, fig.height = 5,
                      warning = F, message = F)
```



```{r}
library(fs)
library(purrr)
library(here)
purrr::walk(.x = fs::dir_ls(here("R")), .f = source)
```

```{r}
dft_drug_feas_sum <- readr::read_rds(
  here('data', 'clin_data_cohort',
       'drug_feas_surv_dmet_sum.rds')
)

dft_drug_feas <- readr::read_rds(
  here('data', 'clin_data_cohort', 
       'drug_feas_surv.rds')
)

dft_drug_surv_nest <- readr::read_rds(
  here(
    'data', 'survival', 'drug', 'fit_outputs', 
    'processed_dmet_drug_nest.rds'
  )
)

```

```{r}
# A little helper function to extract a particular row/column from my nested
#  tibble dataframe structure.
tib_extract <- function(dat, id, col, id_col = "class_comp") {
  if (!(col %in% colnames(dat))) {
    cli_abort("col is not in dat.")
  }
  if (!(id %in% dat[[id_col]])) {
    cli_abort("invalid input for class_comp")
  }
  
  obj <- dat %>%
    filter(.data[[id_col]] %in% id) %>%
    pull(.data[[col]])
  
  obj <- obj[[1]]
  
  return(obj)
  
}


```



## Required conditions

Our goal is analyzing gene features that are predictive biomarkers for select drug classes in the population of metastatic cancer patients.  A bit of notation to make our conditions for inclusion more clear:

- M: The time from diagnosis to diagnosis of distant metastasis.
- D: The first use of the drug class in question for the participant.
- X: The time of the first next-generation sequencing test result for the participant (aka truncation/entry time in our previous analyses).
- Y: The censoring or event time for the event of interest (PFS I+M or OS).

I am considering the following conditions required for a record to be included in the analysis:

- **M≤D**.  Metastatic diagnosis happens before/on the first use of the drug (and they must have a metastatic diagnosis, of course).
- **X≤D**.  The drug start must be after/on the date of the sequencing test result (otherwise the drug causing the genomic result is a plausible explanation).
- **X≤Y**.  The sequencing test must be back before death/event/censoring.  Seems obvious, but not always true in GENIE BPC.

Our key first question is how many records that leaves us with for analysis.

## Eligible cohort sizes

The following table states the number of participants we have who took the drug at any point in the metastatic cohort (n_used_agent) and the number who could be included in an analysis of overall survival based on the time requirements outlined above (n_os).

```{r, include = T}
dft_drug_feas_sum %>%
  rename(` ` = class_comp) %>%
  flextable(.) %>%
  autofit(.)
```

Observations:

- Some of these drug classes are too small to do a robust analysis on.
- The number of people who used a drug is poorly related to the number who could be included in an analysis.  For example about 5% of people who took endocrine therapy would qualify but ~70% of those on IC inhibitors would.
- **Generalizability:  In all drug classes above our cohort is a selective subset of those who took the drug**.  We should be clear about this in the paper.



```{r}
dfp_pfs_feas_sum <- dft_drug_feas %>% 
  filter(had_met & crit_all_os) %>%
  mutate(
    # Use a tolerance of 0.001 yrs = 0.3 days here to avoid rounding erros.
    drug_after_reg = abs(
      drug_dx_start_int_yrs - dx_reg_start_int_yrs
    ) > 0.001,
    prog_before_drug = tt_reg_pfs_i_and_m_dx_yrs < drug_dx_start_int_yrs
  ) %>%
  group_by(class_comp) %>%
  summarize(
    n_elig = n(),
    n_drug_after_reg = sum(drug_after_reg),
    n_prog_before_drug = sum(prog_before_drug),
    n_problem = sum(prog_before_drug & drug_after_reg)
  )
```




```{r, eval = F}

```


## PFS feasibility

Our dataset contains derived variables for survival from start of drug regimen for those that have metastases.  Because we're interested in indexing from select drugs, rather than the regimen containing that drug, there is potentially an issue in using these variables for our analyses.  Specifically, if the following conditions are both met for any subjects meeting inclusion criteria for an analysis, we would have a problem:

1. The drug date and regimen date are not the same (drug started sometime mid-regimen).
2. A progression event is noted for the regimen before the drug is started.

In this case we would not have access to the time from drug start to progression without recalculation of the variable.  Removing participants who meet these criteria from the analysis has an obvious bias potential by filtering out people with systematically low progression times.

We will start by simply calculating the number of people who meet these criteria (`n_problem`):

```{r, include = T}
dfp_pfs_feas_sum %>%
  rename(` ` = class_comp) %>%
  flextable(.) %>%
  autofit(.)
```

**Observations:**

- For all drugs except CDK inhibitors and IC inhibitors there are no people who meet both criteria.  Therefore, we should be fine to use the regimen progression variables (recalculated from drug start) instead of doing the substanial work & guesswork involved in recalculating progression events and times.  
- I will take this substantially easier, slightly incorrect, course in the analysis below.


*Sidenote:*  Two long term fixes for this issue, to allow calculation of PFS from the start of a particular drug, include (1) adding PFS variables for each drug or (2) making the code used to gather PFS events available.  The latter has the advantage of extending to other types of events.



## Results

Here we index from the time of drug initiation for the first use of a drug in a given class.  I have only generated results for drug classes where there at least 30 eligible cases.  This is an arbitrary threshold.  However, inferring anything about predictive biomarkers or using the bootstrap with smaller sizes than this is unwise. but inferring anything from sizes smaller than that is far fetched.

We're applying similar methods here as we did with prognostic variables from the diagnosis of distant metastasis:  Bootstrapping a regularized cox model, and calling the number of bootstrap resamples where a covariate was selected 'stability'.  The main difference here is we don't have left truncation, because we index from drug initiation and filter our cohort based on the additional criteria above.

Adjustment variables include:

- `age_drug_start` - Participant age at drug start (approximate since we only have integer age at diagnosis).
- `stage_dx_iv_num` - Whether a participant was stage IV at diagnosis (1), or developed metastases sometime after diagnosis (0).
- `birth_year_c` - Birth year.
- `white` - Indicator for primary race being white.
- `hispanic` - Indicator for whether the person was hispanic or latino.
- {`bca_her2_pos`, `bca_trip_neg`, `bca_nc_nr`} - Indicators for breast cancer subtype, relative to the reference group HR+/HER2-.  `bca_nc_nr` is "not complete or not relevant", i.e. missing.


```{r}
# Create the tables that state the agents in each drug class:
ft_agent_cdk <- dft_drug_surv_nest %>%
  tib_extract(., "CDK inhibitor", "dat_surv") %>%
  get_agent_ft(.)
```


```{r}
# Create combined plots.  The individual plots were already rendered in the
#   processing step.
gg_pfs_os_cdk <- plot_grid(
  tib_extract(dft_drug_surv_nest, "CDK inhibitor", col = "plot_os_scatter"),
  tib_extract(dft_drug_surv_nest, "CDK inhibitor", col = "plot_pfs_scatter"),
  ncol = 1
)

gg_pfs_os_ic <- plot_grid(
  tib_extract(dft_drug_surv_nest, "IC inhibitor", col = "plot_os_scatter"),
  tib_extract(dft_drug_surv_nest, "IC inhibitor", col = "plot_pfs_scatter"),
  ncol = 1
)

gg_pfs_os_chemo <- plot_grid(
  tib_extract(dft_drug_surv_nest, "chemo", col = "plot_os_scatter"),
  tib_extract(dft_drug_surv_nest, "chemo", col = "plot_pfs_scatter"),
  ncol = 1
)

gg_pfs_os_endo <- plot_grid(
  tib_extract(dft_drug_surv_nest, "endocrine", col = "plot_os_scatter"),
  tib_extract(dft_drug_surv_nest, "endocrine", col = "plot_pfs_scatter"),
  ncol = 1
)

gg_pfs_os_pi3k <- plot_grid(
  tib_extract(dft_drug_surv_nest, "pi3k pathway inhibitor", col = "plot_os_scatter"),
  tib_extract(dft_drug_surv_nest, "pi3k pathway inhibitor", col = "plot_pfs_scatter"),
  ncol = 1
)
```

### CDK inhibitors

Cases which met the eligibility criteria had the following agents:

```{r, include = T}
tib_extract(dft_drug_surv_nest, "CDK inhibitor", "ft_agent")
```


```{r, include = T, fig.height = 7}
gg_pfs_os_cdk
```


```{r, include = T, fig.height = 4}
tib_extract(dft_drug_surv_nest, "CDK inhibitor", "plot_mod_comp")
```


### IC inhibitors

Cases which met the eligibility criteria had the following agents:

```{r, include = T}
tib_extract(dft_drug_surv_nest, "IC inhibitor", "ft_agent")
```


```{r, include = T, fig.height = 7}
gg_pfs_os_ic
```


```{r, include = T, fig.height = 4}
tib_extract(dft_drug_surv_nest, "IC inhibitor", "plot_mod_comp")
```


### Chemotherapies

Cases which met the eligibility criteria had the following agents:

```{r, include = T}
tib_extract(dft_drug_surv_nest, "chemo", "ft_agent")
```


```{r, include = T, fig.height = 7}
gg_pfs_os_chemo
```



```{r, include = T, fig.height = 4}
tib_extract(dft_drug_surv_nest, "chemo", "plot_mod_comp")
```

### Endocrine

Cases which met the eligibility criteria had the following agents:

```{r, include = T}
tib_extract(dft_drug_surv_nest, "endocrine", "ft_agent")
```


```{r, include = T, fig.height = 7}
gg_pfs_os_endo  
```


```{r, include = T, fig.height = 4}
tib_extract(dft_drug_surv_nest, "endocrine", "plot_mod_comp")
```

### PI3K pathway inhibitors

Cases which met the eligibility criteria had the following agents:

```{r, include = T}
tib_extract(dft_drug_surv_nest, "pi3k pathway inhibitor", "ft_agent")
```


```{r, include = T, fig.height = 7}
gg_pfs_os_pi3k
```


```{r, include = T, fig.height = 4}
tib_extract(dft_drug_surv_nest, "pi3k pathway inhibitor", "plot_mod_comp")
```


```{r, eval = F}
# looking into something from last week:
dft_drug_feas %>%
  filter(class_comp %in% "endocrine") %>%
  summarize(
    across(
      .cols = c(contains("crit_"), had_met),
      .fns = function(x) {
        d <- length(x)
        n <- sum(x, na.rm = T)
        return(paste0(n, "/", d, "(", round(n/d*100, 0), "%)"))
      }
    )
  ) %>%
  glimpse

```

