---
title: "Predictive Survival"
subtitle: "Metastatic cohort, drugs grouped by class"
author: "Alex Paynter"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, include = F,
                      fig.width = 7, fig.height = 5,
                      warning = F, message = F)
```



```{r}
library(fs)
library(purrr)
library(here)
purrr::walk(.x = fs::dir_ls(here("R")), .f = source)
```

```{r}
dft_drug_feas_sum <- readr::read_rds(
  here('data', 'clin_data_cohort',
       'drug_feas_surv_dmet_sum.rds')
)

dft_drug_feas <- readr::read_rds(
  here('data', 'clin_data_cohort', 
       'drug_feas_surv.rds')
)

```


## Required conditions

Our goal is analyzing gene features that are predictive biomarkers for select drug classes in the population of metastatic cancer patients.  To make my life easier I will define a few time variables:

- M: The time from diagnosis to diagnosis of distant metastasis.
- D: The first use of the drug class in question for the participant.
- X: The time of the first next-generation sequencing test result for the participant (aka truncation time).
- Y: The censoring or event time for the event of interest (PFS I+M or OS).

I am considering the following conditions required for a record to be included in the analysis:

- **M≤D**.  Metastatic diagnosis happens before/on the first use of the drug.
- **X≤D**.  The drug start must be after/on the date of the sequencing test result (otherwise the drug causing the genomic result is a plausible explanation).
- **X≤Y**.  The sequencing test must be back before death/event/censoring - not always true in GENIE.

Our key first question is how many records that leaves us with for analysis.

## Eligible cohort sizes

The following table states the number of participants we have who took the drug at any point in the metastatic cohort (n_used_agent) and the number who could be included in an analysis of overall survival based on the time requirements outlined above (n_os).

```{r, include = T}
dft_drug_feas_sum %>%
  rename(` ` = class_comp) %>%
  flextable(.) %>%
  autofit(.)
```

Observations:

- Some of these drug classes are probably too small to do a robust analysis on.
- The number of people who used a drug is poorly related to the number who could be included in an analysis.

**PFS note:** PFS is complicated, because some people did not start the drug at the same time as the regimen started, so the derived variables in the datasets may not be sufficient.  However, we can say for sure that `n_pfs` will never be greater than `n_os` because $Y_{pfs} <= Y_{os}$ by definition.

## Example analysis

Here we index from the time of first drug initiation.  Originally the plan was to look at anti-HER2 therapy, but as we can see from the table above the group that's eligible for anti-HER2 therapy is small.  I decided to use CDK inhibitors as a proof-of-concept instead.

Essentially we're applying similar methods here as we did with prognostic variables from the diagnosis of distant metastasis:  Bootstrapping a regularized cox model, and calling the number of bootstrap resamples where a covariate was selected 'stability'.  The main difference here is we don't have left truncation, because we index from drug iniation and filter our cohort based on the additional criteria above.

As a proof of concept, here are the results for CDK inhibitors (no clinical adjustments to start - need to think about these):

```{r}
dft_coef_cdk <- readr::read_rds(
  here('data', 'survival', 'drug', 'fit_outputs', 'fit_summary', 'coef_cdk.rds')
)
```

```{r, include = T}
plot_hr_rel_scatter_gene_clin(
  dft_coef_cdk,
  plot_title = "Varibles predictive of OS from first initiation of CDK inhibitors",
  plot_subtitle = "Limitations: metastatic patients whose first drug use is after at least one sequencing test."
)
```


